Core Data: Fixes necessary for converting to TypeScript

dmsnell opened the pull request:
In this commit we're cleaning up type issues in the core-data package
that prevent us from telling TypeScript to run on the package and all
of its existing code, even the JS files.

After these changes we should be able to do so and start converting
more modules to TypeScript with less friction.

This patch follows a series of other smaller updates:
- #39212
- #39214
- #39225
- #39476
- #39477
- #39479
- #39480
- #39525
- #39526
- #39655
- #39656
- #39659

It was built in order to support ongoing work to add types to the
`getEntityRecord` family of functions in #39025.


## Testing

As a type-only change this shouldn't impact the generated bundles.
Please review the types and the fixes.

Note that the purpose isn't to provide helpful and long-term type
signatures but enable TypeScript support across the entire package.

adamziel commented:
This is closer and closer @dmsnell, great work! I rebased on top of latest `trunk` to make the list of changes reflect the current state.

dmsnell committed:
Core Data: Fixes necessary for converting to TypeScript

In this commit we're cleaning up type issues in the core-data package
that prevent us from telling TypeScript to run on the package and all
of its existing code, even the JS files.

After these changes we should be able to do so and start converting
more modules to TypeScript with less friction.

This patch follows a series of other smaller updates:
 - #39212
 - #39214
 - #39225
 - #39476
 - #39477
 - #39479
 - #39480
 - #39525
 - #39526
 - #39655
 - #39656
 - #39659

It was built in order to support ongoing work to add types to the
`getEntityRecord` family of functions in #39025.

dmsnell committed:
Don't actually enable TypeScript for the core-data package yet

adamziel reviewed (approved):
This megabranch is not so mega anymore. I'd be happy with merging the remaining changes as they are without splitting this any further. Great work @dmsnell!
